kernel/comps/virtio/src/device/block/device.rs

```diff
@@ -186,31 +185,40 @@ impl DeviceInner {
             device_id: None,
         };
 
+        fn config_space_change(_: &TrapFrame) {
+            info!("Virtio block device config space change");
+        }
+        fn dummy_handle_block_device(_: &TrapFrame) {
+            info!("Virtio block device handle block device");
+        }
+        device
+            .transport
+            .register_cfg_callback(Box::new(config_space_change))
+            .unwrap();
+        device
+            .transport
+            .register_queue_callback(0, Box::new(dummy_handle_block_device), false)
+            .unwrap();
+
+        // Sets the queue callback function properly before calling `get_id`;
+        // otherwise, the interrupt mechanism cannot work.
         let device_id = device.get_id();
-        let cloned_device_id = device_id.clone();
 
+        // Assigns the device ID and resets the queue callback function
+        let cloned_device_id = device_id.clone();
+        device.device_id = Some(cloned_device_id);
         let handle_block_device = move |_: &TrapFrame| {
             aster_block::get_device(device_id.as_str())
                 .unwrap()
                 .handle_irq();
         };
-
-        device.device_id = Some(cloned_device_id);
-
-        device
-            .transport
-            .register_cfg_callback(Box::new(config_space_change))
-            .unwrap();
-
         device
             .transport
             .register_queue_callback(0, Box::new(handle_block_device), false)
```

Comment:
The device ID is required before registering this callback because the `handle_block_device` uses the ID internally.

```rust
        let handle_block_device = move |_: &TrapFrame| {
            aster_block::get_device(device_id.as_str())
                .unwrap()
                .handle_irq();
        };
```

But the `handle_irq` method is implemented by the virtio-block driver.

```rust
impl aster_block::BlockDevice for BlockDevice {
    fn enqueue(&self, bio: SubmittedBio) -> Result<(), BioEnqueueError> {
        self.queue.enqueue(bio)
    }

    fn handle_irq(&self) {
        info!("Virtio block device handle irq");
        self.device.do_handle_irq();
    }
}
```

In other words, the driver of a block device---in this case, the virtio-block driver---has all the needed information to process the IRQ. It does not need to rely on the block component to process the IRQ. As such, requesting the device ID before registering the IRQ callback is unnecessary. I think this issue has been identified by @yingdi-shan. And the comment is marked resolved. But I don't think this is resolved.

A simple fix is shown below.

```rust
impl DeviceInner {
    // The return type is modified to return `Arc<Self>` on success.
    fn init(mut transport: Box<dyn VirtioTransport>) -> Result<Arc<Self>, VirtioDeviceError> {
        // ...

        let mut device = Arc::new(Self {
            // ...
        });

        // === Removed ===
        // device
        //    .transport
        //    .register_queue_callback(0, Box::new(dummy_handle_block_device), false)
        //    .unwrap();
        // let device_id = device.get_id();
        // === Removed ===

        let handle_block_device = {
            let cloned = device.clone();
            move |_: &TrapFrame| {
                cloned.handle_irq();
            };
        };
        device
            .transport
            .register_queue_callback(0, Box::new(handle_block_device), false)
            .unwrap();

        Ok(device)
    }
}

```

The new version simplifies the `init` method. More importantly, it avoids looking the block device table every time a IRQ is fired. So the new version also improve efficiency.

Reply:
This gives me a good hint; in fact, the device id does not need to be stored in the device information. Following this idea, I have rewritten this part of the logic.